### ISSUE TITLE
Item_type of inherited classes

### ISSUE THREAD

**Issue by yavora** [airblade/paper_trail#362]
And here i am again, sorry :D

Look, for example, i have something like this:

``` rb
class Parent < ActiveRecord::Base
end

class ChildA < Parent
  has_paper_trail
end

class ChildB < Parent
  has_paper_trail
end
```

So i want to have item_type be equal to 'ChildA' or 'ChildB', not 'Parent', when i'm doing:

``` rb
child_a = ChildA.create(name: 'Name')
child_a.version.item_type #Parent
```

Is it possible? Or am i doing it wrong? 

Thank you.

**Comment by alexpooley** [comment]
I got stung with this issue just now. Paper Trail makes the assumption that inheritance is used explicitly for STI. As referenced in this thread, inheritance can be used for more than just STI such as extending a Gem model. `reify` fails because there is no version_class_name on the parent. A note in the readme would go a long way to saving many wasted hours.

I would add the note myself, but ironically I've just wasted many hours working through the issue. The note would go something like...

----

If you're using model inheritance for something other than STI you will need to override the base_class of the parent class like so:

```
class Tag < ActsAsTaggableOn::Tag
  has_paper_trail

  def self.base_class
    self
  end
end
```

**Comment by laranicolas** [comment]
@alexpooley @Envek I am having the same problem you have mentioned but the solution you provided is not working for me. 

I am using rails v5.2.3, ruby 2.6.3.

Any help?

**Comment by alexpooley** [comment]
Hola @laranicolas , post some code and I can try to help?

**Comment by laranicolas** [comment]
Hola @alexpooley, thanks for answer.

The problem I had was that when I created a VendorLogo, "PaperTrail" create a version using `Upload` as `item_type` instead of using `VendorLogo`.

```
class Upload < ActiveRecord::Base
  has_paper_trail versions: {class_name: 'Moderation::AdPartner::File::Version'}
```

```
class VendorLogo < Upload
  has_paper_trail versions: {class_name: 'Moderation::AdPartner::Logo::Version'}
```

But now I found a solution adding `meta` to `VendorLogo` relation, means each time I created a `VendorLogo` it creates a version using `VendorLogo` as `item_type`.

```
class VendorLogo < Upload
  has_paper_trail versions: {class_name: 'Moderation::AdPartner::Logo::Version'}, meta: { item_type: 'VendorLogo' }
```

Now I have to solve a new issue, I need to set up the `versions` relation to bring using `VendorLogo` types.

Example of issue:

```Moderation::AdPartner::Logo::Version Load (0.8ms)  SELECT  "versions".* FROM "versions" WHERE "versions"."item_id" = $1 AND "versions"."item_type" = $2 ORDER BY id desc LIMIT $3  [["item_id", "141620"], ["item_type", "Upload"], ["LIMIT", 11]]```

What I am looking for:

```
Moderation::AdPartner::Logo::Version Load (0.8ms)  SELECT  "versions".* FROM "versions" WHERE "versions"."item_id" = $1 AND "versions"."item_type" = $2 ORDER BY id desc LIMIT $3  [["item_id", "141620"], ["item_type", "VendorLogo"], ["LIMIT", 11]]
```

**Comment by alexpooley** [comment]
It has been a long time since I looked at this problem so I am only going my instinct. I think meta will not work well as you are discovering because it doesn't change the fundamental type.

What happens when you add 

```
  def self.base_class
    self
  end
```

to the VendorLogo class?

**Comment by laranicolas** [comment]
Raise this error:

```
Traceback (most recent call last):
        1: from (irb):123
ActiveRecord::StatementInvalid (PG::UndefinedTable: ERROR:  relation "vendor_logos" does not exist)
LINE 8:                WHERE a.attrelid = '"vendor_logos"'::regclass
                                          ^
:               SELECT a.attname, format_type(a.atttypid, a.atttypmod),
                     pg_get_expr(d.adbin, d.adrelid), a.attnotnull, a.atttypid, a.atttypmod,
                     c.collname, col_description(a.attrelid, a.attnum) AS comment
                FROM pg_attribute a
                LEFT JOIN pg_attrdef d ON a.attrelid = d.adrelid AND a.attnum = d.adnum
                LEFT JOIN pg_type t ON a.atttypid = t.oid
                LEFT JOIN pg_collation c ON a.attcollation = c.oid AND a.attcollation <> t.typcollation
               WHERE a.attrelid = '"vendor_logos"'::regclass
                 AND a.attnum > 0 AND NOT a.attisdropped
               ORDER BY a.attnum
```

**Comment by alexpooley** [comment]
ok, what is the Moderation::AdPartner::File::Version class ?

**Comment by laranicolas** [comment]
I have to different versions model for `VendorLogo` and `Upload` models:

`Upload` version file:
```
class Moderation::AdPartner::File::Version < PaperTrail::Version
  default_scope { where(item_type: 'Upload') }
  belongs_to :upload, foreign_key: :item_id
  belongs_to :user, foreign_key: :whodunnit, optional: true
end
```

`VendorLogo` version file:
```
class Moderation::AdPartner::Logo::Version < PaperTrail::Version
  default_scope { where(item_type: 'VendorLogo') }
  belongs_to :vendor_logo, foreign_key: :item_id
  belongs_to :user, foreign_key: :whodunnit, optional: true
end
```

**Comment by alexpooley** [comment]
@laranicolas I misunderstood what you were trying to achieve. You have STI + custom version classes. I have not used custom version classes before. From memory I think the `item_type` issue you are seeing in your logs is the same problem matches what I saw in my original issue.

Here's some things I would try if you're out of ideas:

1. Maybe use polymorphic relationship in your PaperTrail::Version subclasses instead of the default_scope + belongs_to
2. Try `def self.base_class; VendorLogo; end` in Moderation::AdPartner::Logo::Version
3. Debug the gem with byebug or `puts` (like the real pros). That's how I found my issue and it was easier than I thought it was going to be. 

Sorry I can't help more!
